tiledb/api/src/array/enumeration/mod.rs

```diff
@@ -0,0 +1,529 @@
+            })
+        } else {
+            Err(self.context.expect_last_error())
+        }
+    }
+}
+
+impl<'ctx> Debug for Enumeration<'ctx> {
+    fn fmt(&self, f: &mut Formatter) -> FmtResult {
+        let name = self.name().map_err(|_| fmt::Error)?;
+        let dtype = self.datatype().map_err(|_| fmt::Error)?;
+
+        let dtype_string =
+            dtype.to_string().unwrap_or("<unknown datatype>".to_owned());
+        let cell_val_num = self.cell_val_num().map_err(|_| fmt::Error)?;
+        let ordered = self.ordered().map_err(|_| fmt::Error)?;
+
+        let json = json!({
+            "name": name,
+            "datatype": dtype_string,
+            "cell_val_num": cell_val_num,
+            "ordered": ordered,
+            "values": [], // TODO: Render values
+        });
+        write!(f, "{}", json)
+    }
+}
+
+impl<'c1, 'c2> PartialEq<Enumeration<'c2>> for Enumeration<'c1> {
+    fn eq(&self, other: &Enumeration<'c2>) -> bool {
+        let names_match = match (self.name(), other.name()) {
+            (Ok(mine), Ok(theirs)) => mine == theirs,
+            _ => false,
+        };
+        if !names_match {
+            return false;
+        }
+
+        let types_match = match (self.datatype(), other.datatype()) {
+            (Ok(mine), Ok(theirs)) => mine == theirs,
+            _ => false,
+        };
+        if !types_match {
+            return false;
+        }
+
+        let cell_val_num_match =
+            match (self.cell_val_num(), other.cell_val_num()) {
+                (Ok(mine), Ok(theirs)) => mine == theirs,
+                _ => false,
+            };
+        if !cell_val_num_match {
+            return false;
+        }
+
+        let ordered_match = match (self.ordered(), other.ordered()) {
+            (Ok(mine), Ok(theirs)) => mine == theirs,
+            _ => false,
+        };
+        if !ordered_match {
+            return false;
+        }
+
+        let data_match = match (self.data(), other.data()) {
+            (Ok(mine), Ok(theirs)) => mine == theirs,
+            _ => false,
+        };
+        if !data_match {
+            return false;
+        }
+
+        let offsets_match = match (self.offsets(), other.offsets()) {
+            (Ok(Some(mine)), Ok(Some(theirs))) => mine == theirs,
+            (Ok(None), Ok(None)) => true,
+            _ => false,
+        };
+        if !offsets_match {
+            return false;
+        }
+
+        true
+    }
+}
+
+pub struct Builder<'ctx> {
+    context: &'ctx Context,
+    name: String,
+    dtype: Datatype,
+    cell_val_num: u32,
+    ordered: bool,
+}
+
+impl<'ctx> Builder<'ctx> {
+    pub fn new(context: &'ctx Context, name: &str, dtype: Datatype) -> Self {
+        Builder {
+            context,
+            name: name.to_owned(),
+            dtype,
+            cell_val_num: 1,
+            ordered: false,
+        }
+    }
+
+    pub fn cell_val_num(self, cell_val_num: u32) -> Self {
```

Comment:
This has to be nonzero right? Maybe (in another PR) we should use the `Nonzero` struct here.

Reply:
Oooh! Nice find. I don't know of any place where a cell_val_num of zero is a thing so that would work nicely.

Reply:
I created sc#44469